### PR TITLE
fix: Use strategies for workspace plugins.

### LINKED ISSUE #1978: Dependency version bumps using the node-workspace plugin to not update version numbers for "extra-files"
#### Environment details

  - OS: Ubuntu 22.04.2
  - Node: 16
  - `release-please` version: 15.10.4

#### Steps to reproduce

1. Using a mono repo with multiple node packages, and with dependencies between those packages. Repo in question: https://github.com/launchdarkly/js-core
2. A package, that is dependent on another package, contains files which are specified in "extra-files" that require version updates. File: https://github.com/launchdarkly/js-core/blob/cdc0acc6cb73eeb98baf7335e950ebc00f4f841f/packages/sdk/vercel/src/createPlatformInfo.ts#L13, Configuration: https://github.com/launchdarkly/js-core/blob/cdc0acc6cb73eeb98baf7335e950ebc00f4f841f/release-please-config.json#L16
3. Update a package that another package depends on.

When the version is bumped by the node-workspaces plugin it seems to only be bumped in the package.json and not in any files specified in the "extra-files". Release PR that demonstrates this: https://github.com/launchdarkly/js-core/pull/147

### PR DESCRIPTION
**Pull request by kinyoklion**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/release-please/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1978,  #2089 and maybe #2109 🦕

The main goal here is to generate a base pull request using the strategy defined for the release. Then add to that the updates that are needed specifically to the workspace. This will result in generating the correct changelog header as well as updating "extra-files".

I need to implement tests for this yet, but I wanted to get some opinions on the approach before piling too much code on. I've only made changes for the node workspace, but the concept could be extended to to the other workspace plugins.

Example pr created with these changes showing fixes for #1978 #2089:
https://github.com/kinyoklion/release-please-monorepo-tests/pull/10/files

### CODE REVIEW

File: src/plugins/node-workspace.ts [hunk 1]
```diff
@@ -271,20 +278,31 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
     );
     const packageJson = updatedPackage.toJSON() as PackageJson;
     const version = Version.parse(packageJson.version);
+
+    const strategy = this.strategiesByPath[updatedPackage.location];
+    const latestRelease = this.releasesByPath[updatedPackage.location];
+
+    const basePullRequest = strategy
```
Comment (kinyoklion) [inline src/plugins/node-workspace.ts:285]: Make a base pull request that contains the "extra-files" updaters as well as the changelog header.

File: src/plugins/node-workspace.ts [hunk 2]
```diff
@@ -297,16 +315,17 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
           createIfMissing: false,
           updater: new Changelog({
             version,
-            changelogEntry: appendDependenciesSectionToChangelog(
-              '',
-              dependencyNotes,
-              this.logger
-            ),
+            versionHeaderRegex: `\n###? v?[${version
```
Comment (kinyoklion) [inline src/plugins/node-workspace.ts:318]: The base PR will include the version. The default regex would insert changes above the last version. Which makes it backward.

I need to come back and cleanup the "append" updater if the PR moves forward.

File: src/strategies/base.ts [hunk 3]
```diff
@@ -254,19 +254,19 @@ export abstract class BaseStrategy implements Strategy {
     commits: ConventionalCommit[],
     latestRelease?: Release,
     draft?: boolean,
-    labels: string[] = []
+    labels: string[] = [],
+    bumpOnlyOptions?: BumpReleaseOptions
```
Comment (kinyoklion) [inline src/strategies/base.ts:258]: I needed some indicator to bypass all the early out conditions associated with no commits or changelog entries.

An alternate approach to this whole thing may be to generate all the PRs always, and then prune ones we don't want.

File: src/plugins/node-workspace.ts [hunk 4]
```diff
@@ -297,16 +315,17 @@ export class NodeWorkspace extends WorkspacePlugin<Package> {
...
+              .replace('.', '.')}]`,
+            changelogEntry: dependencyNotes,
           }),
         },
       ],
       labels: [],
-      headRefName: BranchName.ofTargetBranch(this.targetBranch).toString(),
+      headRefName:
+        basePullRequest?.headRefName ??
+        BranchName.ofTargetBranch(this.targetBranch).toString(),
```
Comment (kinyoklion) [inline src/plugins/node-workspace.ts:328]: For #2109

**Comment by kinyoklion** [comment]
I am working on the CLA situation.

File: src/strategy.ts [hunk 5]
```diff
@@ -47,7 +51,8 @@ export interface Strategy {
     commits: Commit[],
     latestRelease?: Release,
     draft?: boolean,
-    labels?: string[]
+    labels?: string[],
+    bumpOnlyOptions?: BumpReleaseOptions
```
Comment (chingor13) [inline src/strategy.ts:55]: I'm a bit unclear on what these options are for. Perhaps a docstring is good enough to describe what it's use is for.

**Review by chingor13 — CHANGES_REQUESTED** [review]
Thank you for this and apologies for the late review.

In general I think this is a fine approach, although I don't fully understand the edge cases you are trying to handle.  Test case would demonstrate the primary scenario it's fixing and prevent regressions in the future. Test cases would also demonstrate the edge cases you're trying to handle.

**Comment by kinyoklion** [comment]
> Thank you for this and apologies for the late review.
> 
> In general I think this is a fine approach, although I don't fully understand the edge cases you are trying to handle. Test case would demonstrate the primary scenario it's fixing and prevent regressions in the future. Test cases would also demonstrate the edge cases you're trying to handle.

Thank you @chingor13,

Now that I know the general approach is fine I will add some test cases that make this more clear.

The situation is not really an edge case. If you have any workspace dependent packages, then it doesn't generate change logs correctly or update files with version numbers.

Imagine you have two packages. Package A and package B. Package B depends on package A.

You commit a change which affects files in package A: "feat: Add my feature."

The release PR that gets generated will have an incorrect changelog. It will just append a "dependencies" section to the top of the changelog.

Example:
image

The second major problem is that if package B had any `extra-files`, then the version numbers in those files would not be updated.

Both of these are because version bumps only from dependencies do not run any of the steps associated with the base strategy being used.

Thank you,
Ryan

File: src/strategy.ts (hunk 5 above)
Comment (kinyoklion) [reply src/strategy.ts:55]: Yes, I will add some documentation. 

I wasn't 100% sure on this, but basically these serve as an indicator that you are creating a release with a version number update, even if there were no associated conventional commits. 

So it is:
A: An indicator that we want to make a release even though there are no calculated changes.
B: The version number we want for that release.

File: __snapshots__/node-workspace.js [hunk 6]
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
...
+
+
+<details><summary>@here/pkgA: 3.3.4</summary>
+
+Release notes for path: node1, releaseType: node
+</details>
+
+<details><summary>@here/pkgB: 2.2.3</summary>
+
+## [2.2.3](https://github.com/googleapis/node-test-repo/compare/pkgB-v2.2.2...pkgB-v2.2.3) (1983-10-10)
```
Comment (kinyoklion) [inline __snapshots__/node-workspace.js:153]: This line here represents one of the primary differences. Without this change, there would be no header line here. Both in the PR and in the Changelog.

File: __snapshots__/node-workspace.js [hunk 7]
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
...
+
+---
+This PR was generated with [Release Please](https://github.com/googleapis/release-please). See [documentation](https://github.com/googleapis/release-please#release-please).
+`
+
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 2'] = `
+other notes
+`
+
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 3'] = `
```
Comment (kinyoklion) [inline __snapshots__/node-workspace.js:195]: This demonstrates the correct header line being added to the changelog.

File: __snapshots__/node-workspace.js [hunk 8]
```diff
@@ -138,6 +138,84 @@ Release notes for path: node1, releaseType: node
...
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 4'] = `
+### Dependencies
+
+* The following workspace dependencies were updated
+  * dependencies
+    * @here/pkgA bumped from 3.3.3 to 3.3.4
+`
+
+exports['NodeWorkspace plugin run includes headers for packages with configured strategies 5'] = `
+## 1.1.2 (2023-12-15)
```
Comment (kinyoklion) [inline __snapshots__/node-workspace.js:208]: This also would not have been present. This doesn't include a diff, because it didn't have the previous version information included.

File: test/plugins/node-workspace.ts [hunk 9]
```diff
@@ -372,6 +376,171 @@ describe('NodeWorkspace plugin', () => {
...
+
+      const newCandidates = await plugin.run(candidates);
+      expect(newCandidates).lengthOf(1);
+      const nodeCandidate = newCandidates.find(
+        candidate => candidate.config.releaseType === 'node'
+      );
+      expect(nodeCandidate).to.not.be.undefined;
+      const updates = nodeCandidate!.pullRequest.updates;
+
+      assertHasUpdate(updates, 'node2/my-file', Generic);
```
Comment (kinyoklion) [inline test/plugins/node-workspace.ts:542]: Without invoking the base strategy we would not have the update for extra-files.

**Comment by kinyoklion** [comment]
I've added some additional doc comments, added tests, and made some comments in the PR for those tests.

In this PR I am only focusing on node, but I believe that these same problems likely affect the other workspace plugins.

**Comment by kinyoklion** [comment]
I found some situation where the changelog is still not correct, so I am investigating that and will update.

**Comment by TimothyJones** [comment]
I am certain this would fix #2109. 

In addition to the issues you mentioned, I think this _might_ also fix #1946, or at least address one of the issues that contributes to it.

**Review by chingor13 — APPROVED** [review]

**Review by GregKruz — APPROVED** [review]
